File: organism/creatures.py
from __future__ import annotations
from organism.genetics import Genome, CreatureTypes, creatures_genomes, sociability_species
from organism.stats import LimitedValue, Energy, Life, Age, Fertility
from decisions.actions import Intent, IntentActs
from organism.ontology import Gender, AttackedEvent

from core.error import IdNotFoundError, IdAlreadyExistsError
from organism.identity import Id, EntityTypes, gen_id
from random import uniform
from math import exp
from dataclasses import dataclass
from core.coord import Coord
from utils.namegenerator import gen_name
from typing import Iterable



MAX_STRENGTH = 35


class Gestation:
    def __init__(self, limit:int):
        self.value = 0
        self.limit = limit
    @property
    def is_overdue(self) -> bool:
        return self.value > self.limit + 1
    @property
    def is_ready_to_born(self) -> bool:
        return self.value >= self.limit
    @property
    def is_premature(self) -> bool:
        return self.value + 1 < self.limit 
    @property
    def is_at_point(self) -> bool:
        return self.value == self.limit
    @property
    def death_factor(self) -> float: # [0, 1]
        factor = 0

        if self.is_premature:
            factor += self.time_to_born/(self.time_to_born + self.limit)*1.40
        elif self.is_overdue:
            factor += self.overdue/(self.overdue + self.limit)
        else: # at_point
            pass
        return factor

    @property
    def overdue(self) -> int:
        return max(0, self.value - self.limit)
    @property
    def time_to_born(self) -> int:
        return max(0, self.limit - self.value)

    @property
    def ratio(self) -> float:
        return self.value/self.limit
    @property
    def overdue_factor(self) -> float:
        return abs(self.overdue)






@dataclass
class PregnantUterus:
    male_genome:Genome
    gestation:Gestation
    number_children:LimitedValue
    @property
    def all_children_borned(self) -> bool:
        return self.number_children.value == self.number_children.limit 
    
    
    @property
    def pregnancy_cost(self) -> float:
        return (1 + self.gestation.value/2)*self.number_children.value
    @property
    def gravity(self) -> float:
        return ((self.number_children.ratio + self.gestation.ratio)/2)**1.4
    @property
    def birth_energy(self) -> float:
        return self.pregnancy_cost / self.number_children.value 
    
@dataclass(frozen=True)
class EmptyUterus:
    pass





class Creature:
    def __init__(self,
                 genome:Genome,
                 gender:Gender,
                 name:str,
                 initial_energy:Energy,
                 position:Coord,
                 id:str,
                 sociability:int | float
                 ):
        self.genome = genome
        self.gender = gender
        self.name = name
        self.id = Id(id, EntityTypes.CREATURE)

        self.energy = initial_energy
        self.life = Life(value=genome.body.life_limit, limit=genome.body.life_limit)
        self.fertility = Fertility(value=genome.reproduction.fertility_limit, limit=genome.reproduction.fertility_limit)
        self.uterus:EmptyUterus | PregnantUterus | None = EmptyUterus() if self.gender == Gender.FEMALE else None
        self.age = Age(value=0, limit=genome.body.life_limit)

        self.intent: Intent = Intent(IntentActs.NOTHING)
        self.last_attack: None | AttackedEvent = None
        self.position = position

        self.sociability = LimitedValue(sociability, 1, -1)
    @property
    def hungry(self) -> float:
        return 1 - self.energy.ratio
    @property
    def needed_energy(self) -> float:
        return self.energy.limit - self.energy.value

    @property
    def reproductive_maturity(self) -> float:
        return exp(-(self.age.ratio - 0.45)**2/(0.2)**2)
    @property
    def reproductively_capable(self) -> bool:
        energy_condition = self.energy.value * self.genome.reproduction.extra_reproduction_multiplier >= self.genome.reproduction.reproduction_cost
        return self.fertility.reproductive_capability() and not self.pregnant and energy_condition
    @property
    def reproductive_fitness(self) -> float:
        '''
        < 1 -> not so good
        = 1 -> at point
        > 1 -> good
        '''
        return self.energy.value * self.genome.reproduction.extra_reproduction_multiplier  / self.genome.reproduction.reproduction_cost
    
    @property
    def strength_factor(self) -> int | float:
        return self.genome.body.strength / MAX_STRENGTH
    @property
    def physical_ratio(self) -> int | float:
        '''
        Always return a value in [0, 1]
        '''
        return (self.energy.ratio*1.2 + self.life.ratio*1.5 + self.strength_factor*0.9)/3.6
    @property
    def senescence(self) -> float:
        return self.age.value/(self.age.limit**1.4)
    
    @property
    def pregnant(self) -> bool:
        return isinstance(self.uterus, PregnantUterus)
    @property
    def basal_metabolism(self) -> float:
        return self.genome.metabolism.mass

    
    def __str__(self):
        return f'{self.name}'

class Corpse:
    def __init__(self, energy:Energy, id:str, decomposition_time:Age, position:Coord):
        self.energy = energy
        self.id = Id(id, EntityTypes.CORPSE)
        self.decomposition_time = decomposition_time
        self.position = position
    @property
    def time_left(self) -> float:
        return self.decomposition_time.limit - self.decomposition_time.value
    @property
    def ready_to_disapear(self) -> bool:
        return self.decomposition_time.value == self.decomposition_time.limit
    


    



class EntitiesRegistry:
    def __init__(self):
        self.entitys:dict[Id, Corpse | Creature] = {}
    @property
    def entities(self) -> Iterable[Corpse | Creature]:
        return self.entitys.values()
    def add(self, entity:Creature | Corpse) -> str:
        if entity.id in self.entitys:
            raise IdAlreadyExistsError('ID {} exists'.format(entity.id))
        self.entitys[entity.id] = entity
        return entity.id.id
    def delete(self, id:Id) -> None:
        if id not in self.entitys:
            raise IdNotFoundError('ID {} does not exists'.format(id))
        del self.entitys[id]
    
    def get_creature(self, id:Id) -> Creature:
        if id.e_type != EntityTypes.CREATURE:
            raise TypeError('Can not get creature: ID {} type must be CREATURE'.format(id))
        if id not in self.entitys:
            raise IdNotFoundError('ID {} does not exists'.format(id))
        return self.entitys[id] # type: ignore
    
    def get_corpse(self, id:Id) -> Corpse:
        if id.e_type != EntityTypes.CORPSE:
            raise TypeError('Can not get corpse: ID {} type must be CORPSE'.format(id))
        if id not in self.entitys:
            raise IdNotFoundError('ID {} does not exists'.format(id))
        return self.entitys[id] # type: ignore
    
    def get(self, id:Id) -> Creature | Corpse:
        if id not in self.entitys:
            raise IdNotFoundError('ID {} does not exists'.format(id))
        return self.entitys[id]
    
    
class CreatureFactory:
    @staticmethod
    def gen_creature(
        position:Coord | None = None,
        creature_type:CreatureTypes | None = None,
        id:str | None = None,
        initial_energy:Energy | None = None,
        genome:Genome | None = None,
        gender:Gender | None = None,
        name:str | None = None,
        sociability:int | float | None = None
    ) -> Creature:
        
        if creature_type is None:
            creature_type = CreatureTypes.choice()
        if id is None:
            id = gen_id()
        if genome is None:
            genome = creatures_genomes.get_type_genome(creature_type)
        if initial_energy is None:
            initial_energy = Energy(uniform(0.5, 1) * genome.metabolism.energy_limit, genome.metabolism.energy_limit)

        if gender is None:
            gender = Gender.choice()
        if position is None:
            position = Coord(0, 0)
        if name is None:
            name = gen_name()
        if sociability is None:
            sociability = sociability_species[creature_type]
        return Creature(
            genome, 
            gender,
            name,
            initial_energy,
            position,
            id,
            sociability
        )
    

    

File: src/karkinolution/organism/creatures.cpp

#include <karkinolution/organism/ontology.hpp>
#include <cassert>
#include <karkinolution/organism/genetics.hpp>
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <karkinolution/organism/creatures.hpp>
#include <karkinolution/core/stats.hpp>
#include <variant>
#include <karkinolution/utils/namegenerator.hpp>
#include <karkinolution/utils/k_random.hpp>


// Gestation

bool Gestation::is_overdue() const {
    return value() > max() + 1;
}

bool Gestation::is_ready_to_born() const {
    return value() >= max();
}

bool Gestation::is_premature() const {
    return value() + 1 < max();
}

bool Gestation::is_at_point() const {
    return value() == max();
}

NormalizedValue Gestation::death_factor() const {
    NormalizedValue factor = {0};

    if (is_premature()) {
        uint16_t time_born = time_to_born();
        factor += static_cast<float>(time_born)/(time_born + max())*1.40f;
    } else if (is_overdue()) {
        uint16_t overd = overdue();
        factor += static_cast<float>(overd)/(overd + max());
    }

    return factor;
}


uint16_t Gestation::overdue() const {
    return std::max<uint16_t>(0, (value() - max()) ? value() > max() : 0);
}

uint16_t Gestation::time_to_born() const {
    return std::max<uint16_t>(0, (max() - value()) ? max() > value() : 0);
}

// PregnantUterus

bool PregnantUterus::all_children_borned() const {
    return children_count.value() == children_count.max();
}

float PregnantUterus::pregnancy_cost() const {
    return (1 + static_cast<float>(gestation.value())/2)*children_count.value();
}

NormalizedValue PregnantUterus::gravity() const {
    return std::pow(((children_count.ratio().value() + gestation.ratio().value())/2), 1.4);
}

float PregnantUterus::birth_energy() const {
    return pregnancy_cost()/children_count.value();
}

// Creature

NormalizedValue Creature::hungry() const {
    return NormalizedValue{1.0f - energy.ratio().value()};
}

float Creature::needed_energy() const {
    return energy.max() - energy.value();
}

NormalizedValue Creature::reproductive_maturity() const {
    return static_cast<float>(std::exp(std::pow(std::pow(-(-age.ratio().value() - 0.45)
                            , 2)/(0.2), 2)));
}

bool Creature::pregnant() const {
    return std::holds_alternative<PregnantUterus>(uterus);
}

bool Creature::reproductively_capable() const {
    bool energy_condition = energy.value() * genome.reproduction.extra_reproduction_multiplier > genome.reproduction.reproduction_cost;
    return fertility.reproductive_capability() && energy_condition && ! pregnant();
}

NormalizedValue Creature::reproductive_fitness() const {
    return NormalizedValue{energy.value() * genome.reproduction.extra_reproduction_multiplier / genome.reproduction.reproduction_cost};
}

NormalizedValue Creature::strength_factor() const {
    return NormalizedValue{genome.body.strength/MAX_STRENGTH};
}

NormalizedValue Creature::physical_ratio() const {
    return NormalizedValue{static_cast<float>((energy.ratio().value()*1.2 + life.ratio().value()*1.5 + strength_factor().value()*0.9)/3.6)};
}

NormalizedValue Creature::senescence() const {
    return NormalizedValue{static_cast<float>(age.value()/(std::pow(age.max(), 1.4)))};
}

float Creature::basal_metabolism() const {
    return genome.metabolism.mass;
}

// Corpse

int Corpse::time_left() const {
    return decomposition_time.max() - decomposition_time.value() ? decomposition_time.max() > decomposition_time.value() : 0;
}

bool Corpse::ready_to_disapear() const {
    return decomposition_time.value()==decomposition_time.max();
}

// EntitiesRegistry

Creature& EntitiesRegistry::at_creature(Id id) {
    assert(id.entity_type == EntityTypes::CREATURE);
    return std::get<Creature>(at(id));
}

const Creature& EntitiesRegistry::at_creature(Id id) const {
    assert(id.entity_type == EntityTypes::CREATURE);
    return std::get<Creature>(at(id));
}

Corpse& EntitiesRegistry::at_corpse(Id id) {
    assert(id.entity_type == EntityTypes::CORPSE);
    return std::get<Corpse>(at(id));
}

const Corpse& EntitiesRegistry::at_corpse(Id id) const {
    assert(id.entity_type == EntityTypes::CORPSE);
    return std::get<Corpse>(at(id));
}

Id EntityGetters::get_id(const Entity& entity) {
    return std::visit([](const auto& entity) {
        return EntityGetters::get_id(entity);
    }, entity);
}

Id EntityGetters::get_id(const Corpse& corpse) {
    return IDF::create_corpse_id(corpse.id);
}

Id EntityGetters::get_id(const Creature& creature) {
    return IDF::create_creature_id(creature.id);
}


Vec2 EntityGetters::get_position(const Entity& entity) {
    return std::visit([](const auto& entity) {
        return EntityGetters::get_position(entity);
    }, entity);
}

Vec2 EntityGetters::get_position(const Corpse& corpse) {
    return corpse.position;
}

Vec2 EntityGetters::get_position(const Creature& creature) {
    return creature.position;
}
// CreatureFactory

Creature CreatureFactory::gen_creature(const CreatureConfig& config)
{
    auto position = config.position.value_or(Vec2(0, 0));

    auto specie = config.specie.value_or(
        CreatureSpeciesF::choice()
    );

    auto genome = config.genome.value_or(
        global_creature_genomes.get_genome(specie)
    );

    auto id = config.id.value_or(
        IDF::create_creature_id(IDF::gen_id())
    );

    auto initial_energy = config.initial_energy;

    if (!initial_energy.has_value()) {
        float initial_energy_value =
            Disturbs::gen_disturb(0.5, 1) * genome.metabolism.energy_limit;

        initial_energy = Energy(
            initial_energy_value,
            genome.metabolism.energy_limit
        );
    }

    auto gender = config.gender.value_or(
        GenderF::choice()
    );

    auto name = config.name.value_or(
        gen_name()
    );

    auto sociability = config.sociability.value_or(
        sociability_species.at(specie)
    );

    Uterus uterus{std::monostate{}};

    if (gender == Gender::FEMALE) {
        uterus = EmptyUterus{};
    }

    return Creature{
        genome,
        gender,
        std::move(name),
        id.value,
        *initial_energy,
        Life(
            genome.body.life_limit,
            genome.body.life_limit
        ),
        Age{
            0,
            static_cast<uint16_t>(genome.body.max_age)
        },
        Fertility{
            0,
            genome.reproduction.fertility_limit
        },
        std::move(uterus),
        sociability,
        position
    };
}